service/src/main/java/org/cbioportal/service/impl/MutationEnrichmentServiceImpl.java

```diff
@@ -0,0 +1,129 @@
+        List<MutationSampleCountByGene> mutationSampleCountByGeneList = mutationService
+            .getSampleCountByEntrezGeneIdsAndSampleIds(geneticProfileId, allSampleIds, null);
+        mutationSampleCountByGeneList.removeIf(m -> entrezGeneIds.contains(m.getEntrezGeneId()));
+        List<Gene> genes = geneService.fetchGenes(mutationSampleCountByGeneList.stream().map(m -> 
+            String.valueOf(m.getEntrezGeneId())).collect(Collectors.toList()), "ENTREZ_GENE_ID", "SUMMARY");
+
+        Map<Integer, List<Mutation>> mutationMap = mutationService.fetchMutationsInGeneticProfile(geneticProfileId,
+            alteredSampleIds, null, "ID", null, null, null, null).stream().collect(Collectors.groupingBy(
+                Mutation::getEntrezGeneId));
+
+        mutationSampleCountByGeneList.sort(Comparator.comparing(MutationSampleCountByGene::getEntrezGeneId));
+        genes.sort(Comparator.comparing(Gene::getEntrezGeneId));
+        List<AlterationEnrichment> result = new ArrayList<>();
+        for (int i = 0; i < mutationSampleCountByGeneList.size(); i++) {
+            MutationSampleCountByGene mutationSampleCountByGene = mutationSampleCountByGeneList.get(i);
+            result.add(createAlterationEnrichment(mutationMap.get(mutationSampleCountByGene.getEntrezGeneId()), 
+                mutationSampleCountByGene, genes.get(i), alteredSampleIds.size(), unalteredSampleIds.size()));
+        }
+        
+        setQValue(result);
+        return result;
+    }
+
+    private AlterationEnrichment createAlterationEnrichment(List<Mutation> mutations, 
+                                                            MutationSampleCountByGene mutationSampleCountByGene,
+                                                            Gene gene, int alteredSampleCount, 
+                                                            int unalteredSampleCount) {
+        
+        AlterationEnrichment alterationEnrichment = new AlterationEnrichment();
+        
+        if (mutations == null) {
+            alterationEnrichment.setNumberOfSamplesInAlteredGroup(0);
+        } else {
+            alterationEnrichment.setNumberOfSamplesInAlteredGroup(mutations.stream().collect(Collectors.groupingBy(
+                Mutation::getSampleId)).size());
+        }
+        alterationEnrichment.setNumberOfSamplesInUnalteredGroup(mutationSampleCountByGene.getSampleCount() -
+            alterationEnrichment.getNumberOfSamplesInAlteredGroup());
+        alterationEnrichment.setEntrezGeneId(mutationSampleCountByGene.getEntrezGeneId());
+        alterationEnrichment.setHugoGeneSymbol(gene.getHugoGeneSymbol());
+        alterationEnrichment.setCytoband(gene.getCytoband());
+        setLogRatio(alterationEnrichment, alteredSampleCount, unalteredSampleCount);
+        setPValue(alterationEnrichment, alteredSampleCount, unalteredSampleCount);
+        return alterationEnrichment;
+    }
+
+    private void setQValue(List<AlterationEnrichment> alterationEnrichments) {
```

Comment:
I try to avoid `set...` methods if they are not simple setters, `calc...` might be better in those cases.